confirmation_rule.py

```diff
@@ -311,29 +391,20 @@ def get_latest_confirmed(store: Store) -> Root:
     # and already confirmed block might not be safe to use hence revert to the safest one which is the finalized block
     # this reversal trades monotonicity in favour of safety in the casey of asynchrony in the network
     head = get_head(store)
-    is_confirmed_root_canonical = (confirmed_root == get_ancestor(store, head, store.blocks[confirmed_root].slot))
-    if confirmed_block_epoch + 1 < current_epoch or not is_confirmed_root_canonical:
+    if confirmed_block_epoch + 1 < current_epoch or not is_ancestor(store, head, confirmed_root):
         confirmed_root = store.finalized_checkpoint.root
-
-    # use unrealized justified root if the confirmed root is older
-    # and the following conditions are met:
-    #     1) UJ checkpoint is from the current epoch
-    #     2) by the beginning of the previous slot it is known that the previous epoch is justified
-    #     3) previous epoch checkpoint is an ancestor of UJ checkpoint
-    #
-    # useful in the case when the confirmation rule machinery should be restarted after a period of asynchrony
-    # UJ becomes GJ in the beginning of the next epoch
-    # which, assuming synchrony, every honest validator will vote for
+        
+    # if we are at the beginning of the epoch and the epoch of the unrealized justified checkpoint at beginning of the last slot of
+    # the previous epoch corresponds to the previous epoch, then we can confirm the block of the unrealized justified
+    # checkpoint as, under synchrony, such a checkpoint is for sure now the greatest justified checkpoint in the view
+    # of any honest validator and, therefore, any honest validator will keep voting for it for the entire epoch
     confirmed_block_slot = store.blocks[confirmed_root].slot
-    unrealized_justified_slot = store.blocks[store.unrealized_justified_checkpoint.root].slot
-    unrealized_justified_root = store.unrealized_justified_checkpoint.root
-    prev_slot_justified_slot = store.blocks[store.prev_slot_justified_checkpoint.root].slot
-    prev_slot_justified_root = store.prev_slot_justified_checkpoint.root
-    if (confirmed_block_slot < unrealized_justified_slot
-        and store.unrealized_justified_checkpoint.epoch == current_epoch
-        and store.prev_slot_justified_checkpoint.epoch + 1 == current_epoch
-        and prev_slot_justified_root == get_ancestor(store, unrealized_justified_root, prev_slot_justified_slot)):
-            confirmed_root = store.unrealized_justified_checkpoint.root
+    prev_unrealized_justified_checkpoint_slot = store.blocks[store.prev_slot_unrealized_justified_checkpoint.root].slot
+
+    if (get_current_slot(store) % SLOTS_PER_EPOCH == 0
+        and store.prev_slot_unrealized_justified_checkpoint.epoch + 1 == current_epoch 
+        and confirmed_block_slot < prev_unrealized_justified_checkpoint_slot):
```

Comment:
```suggestion
        and store.prev_slot_unrealized_justified_checkpoint.epoch + 1 == current_epoch:
```
The above check doesn’t make sense anymore as it is always true

Reply:
I think that it could very well be that in the previous epoch we confirmed a block $b$ descendant of its checkpoint $C$ and at the end of the epoch we see that $C$ is unrealized justified. In this case, we should not set the confirmed block to $block(C)$.